server/channels/store/searchlayer/layer.go

```diff
@@ -111,6 +111,32 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
 	}
 }
 
+func (s *SearchStore) indexChannelsForTeam(rctx request.CTX, teamID string) {
+	var (
+		page     int
+		perPage  = 100
+		channels []*model.Channel
+	)
+
+	for {
+		cs, err := s.channel.GetPublicChannelsForTeam(teamID, page, perPage)
+		if err != nil {
+			rctx.Logger().Warn("Encountered error while retreiving public channels for indexing", mlog.String("team_id", teamID), mlog.Err(err))
+			return
+		}
+
+		channels = append(channels, cs...)
+
+		if len(channels) < perPage {
+			break
+		}
```

Comment:
This will need reviewing as we should check for `len(cs)` or the loop may never finish if `perPage` elements are returned on the first iteration. See https://hub.mattermost.com/xyz/pl/z5k46zi1m3g67d4adp4xgzye5o

One thought (to avoid missing this in the future) could be to build a small abstraction since this pattern is used in a few places and is likely to be re-used in the future. Example:

```go
func pager[T any](fetch func(page int) ([]T, error), perPage int) ([]T, error) {
	var list []T
	var page int

	for {
		fetched, err := fetch(page)
		if err != nil {
			return nil, err
		}

		list = append(list, fetched...)

		if len(fetched) < perPage {
			break
		}

		page++
	}

	return list, nil
}
```

An additional thought is that we may consider improving our dataset for unit tests since it's fairly common to have many channels, so clearly not an edge case.

Let me know what you think.

Reply:
I love the idea of this function, created something in the `uitls` package, please take a look @streamer45